overlord/ifacestate/ifacemgr.go

```diff
@@ -85,25 +85,43 @@ func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*st
 		Optional: true,
 	}
 	summary := fmt.Sprintf(i18n.G("Prepare connection of plug %s:%s"), plugSnap, plugName)
-	collectPlugAttr := hookstate.HookTask(s, summary, plugHookSetup, nil)
+	preparePlugConnection := hookstate.HookTask(s, summary, plugHookSetup, nil)
 
 	slotHookSetup := &hookstate.HookSetup{
 		Snap:     slotSnap,
 		Hook:     "prepare-slot-" + slotName,
 		Optional: true,
 	}
 	summary = fmt.Sprintf(i18n.G("Prepare connection of slot %s:%s"), slotSnap, slotName)
-	collectSlotAttr := hookstate.HookTask(s, summary, slotHookSetup, nil)
+	prepareSlotConnection := hookstate.HookTask(s, summary, slotHookSetup, nil)
 
 	summary = fmt.Sprintf(i18n.G("Connect %s:%s to %s:%s"),
 		plugSnap, plugName, slotSnap, slotName)
 	connectInterface := s.NewTask("connect", summary)
 	connectInterface.Set("slot", interfaces.SlotRef{Snap: slotSnap, Name: slotName})
 	connectInterface.Set("plug", interfaces.PlugRef{Snap: plugSnap, Name: plugName})
-	connectInterface.WaitFor(collectPlugAttr)
-	connectInterface.WaitFor(collectSlotAttr)
+	confirmPlugHookSetup := &hookstate.HookSetup{
+		Snap:     plugSnap,
+		Hook:     "connect-plug-" + plugName,
+		Optional: true,
+	}
+	summary = fmt.Sprintf(i18n.G("Confirm connection of plug %s:%s"), plugSnap, plugName)
+	confirmPlugConnection := hookstate.HookTask(s, summary, confirmPlugHookSetup, nil)
+
+	confirmSlotHookSetup := &hookstate.HookSetup{
+		Snap:     slotSnap,
+		Hook:     "connect-slot-" + slotName,
+		Optional: true,
+	}
+	summary = fmt.Sprintf(i18n.G("Confirm connection of slot %s:%s"), slotSnap, slotName)
+	confirmSlotConnection := hookstate.HookTask(s, summary, confirmSlotHookSetup, nil)
+
+	prepareSlotConnection.WaitFor(preparePlugConnection)
+	connectInterface.WaitFor(prepareSlotConnection)
+	confirmSlotConnection.WaitFor(connectInterface)
+	confirmPlugConnection.WaitFor(confirmSlotConnection)
 
-	return state.NewTaskSet(collectPlugAttr, collectSlotAttr, connectInterface), nil
+	return state.NewTaskSet(preparePlugConnection, prepareSlotConnection, connectInterface, confirmPlugConnection, confirmSlotConnection), nil
```

Comment:
@pedronis The order of tasks matters and it is enforced by WaitFor-s above (this is the order you mentioned, i.e. preparePlug, prepareSlot, connect, confirmSlot, confirmPlug). The order of arguments of state.NewTaskSet(..) list doesn't matter, however it didn't match that order implied by WaitFor-s and I agree this might be confusing; I've changed it to match. Thanks for spotting.